% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/experiment_plots.R
\name{transcriptWindow}
\alias{transcriptWindow}
\title{Make 100 bases size meta window for all libraries in experiment}
\usage{
transcriptWindow(
  leaders,
  cds,
  trailers,
  df,
  outdir = NULL,
  scores = c("sum", "transcriptNormalized"),
  allTogether = TRUE,
  colors = experiment.colors(df),
  title = "Coverage metaplot",
  windowSize = min(100, min(widthPerGroup(leaders, FALSE)), min(widthPerGroup(cds,
    FALSE)), min(widthPerGroup(trailers, FALSE))),
  returnPlot = is.null(outdir),
  dfr = NULL,
  idName = "",
  plot.ext = ".pdf",
  type = "ofst",
  is.sorted = FALSE,
  drop.zero.dt = TRUE,
  BPPARAM = bpparam()
)
}
\arguments{
\item{leaders}{a \code{\link{GRangesList}} of leaders (5' UTRs)}

\item{cds}{a \code{\link{GRangesList}} of coding sequences}

\item{trailers}{a \code{\link{GRangesList}} of trailers (3' UTRs)}

\item{df}{an ORFik \code{\link{experiment}}}

\item{outdir}{directory to save to (default: NULL, no saving)}

\item{scores}{scoring function (default: c("sum", "transcriptNormalized")),
see ?coverageScorings for possible scores.}

\item{allTogether}{plot all coverage plots in 1 output? (defualt: TRUE)}

\item{colors}{Which colors to use, default auto color from function
\code{\link{experiment.colors}}, new color per library type.
Else assign colors yourself.}

\item{title}{title of ggplot}

\item{windowSize}{size of binned windows, default: 100}

\item{returnPlot}{return plot from function, default is.null(outdir),
so TRUE if outdir is not defined.}

\item{dfr}{an ORFik \code{\link{experiment}} of RNA-seq to
normalize against. Will add RNA normalized to plot name if this is done.}

\item{idName}{A character ID to add to saved name of plot,
if you make several plots in the same folder,
and same experiment, like splitting transcripts in two groups like
targets / nontargets etc. (default: "")}

\item{plot.ext}{character, default: ".pdf". Alternatives: ".png" or ".jpg".}

\item{type}{a character(default: "ofst"), load files in experiment
or some precomputed variant, either "ofst", "pshifted" or "default".
These are made with ORFik:::simpleLibs(), shiftFootprintsByExperiment()..
Will load default if bedoc is not found}

\item{is.sorted}{logical (FALSE), is grl sorted. That is + strand groups in
increasing ranges (1,2,3), and - strand groups in decreasing ranges (3,2,1)}

\item{drop.zero.dt}{logical FALSE, if TRUE and as.data.table is TRUE,
remove all 0 count positions.
This greatly speeds up and most importantly, greatly reduces memory usage.
Will not change any plots, unless 0 positions are used in some sense.
(mean, median, zscore coverage will only scale differently)}

\item{BPPARAM}{how many cores/threads to use? default: bpparam()}
}
\value{
NULL, or ggplot object if returnPlot is TRUE
}
\description{
Gives you binned meta coverage plots, either saved seperatly or
all in one.
}
\examples{
df <- ORFik.template.experiment()[3,] # Only third library
loadRegions(df) # Load leader, cds and trailers as GRangesList
#transcriptWindow(leaders, cds, trailers, df, outdir = "directory/to/save")
}
\seealso{
Other experiment plots: 
\code{\link{transcriptWindow1}()},
\code{\link{transcriptWindowPer}()}
}
\concept{experiment plots}
